README
NAME
    Log::Sigil - show warnings with sigil prefix

SYNOPSIS
      filename: synopsis.pl
       1 use Log::Sigil qw( swarn swarn2 );
       2
       3 sub foo {
       4     swarn( "foo" );
       5     swarn( "bar" );
       6 }
       7
       8 swarn( "foo" );
       9
      10 foo( );
      11
      12 swarn( "bar" );

      above shows:
      === foo by synopsis.pl[8]: main::
      +++ foo by synopsis.pl[4]: main::foo
      +++ bar by synopsis.pl[5]: main::foo
      === bar by synopsis.pl[12]: main::

DESCRIPTION
    This module helps printing debug by adding prefix to the warning
    message. The prefix will change if caller changes, meaning 'foo' sub,
    and 'bar' sub have different prefix each other.

    i do printing debug frequently. In debugging, my warning messages became
    too big to read. When i in trouble (yes, so doing printing debug), i do
    not want to remove the warning messages. Once i thought it is needed, it
    is needed twice, and more. Thus, i need a format which can read warning
    messages even if that is big.

EXPORTS
    swarn
    swarn2

FUNCTIONS
    swarn
        Works all of this module does. That are, adding prefix, setting up
        filename and line, and, setting up package and subroutine.

    swarn2
        Same as swarn, but has a 1 bias. This is useful when calling from
        some handler subroutine, such as;

          ( my $ua = LWP::UserAgent->new )->add_handler(
              request_prepare => sub {
                  my( $req, $ua, $h ) = @_;
                  swarn2( "Adding If-Modified-Since..." );
                  $req->...
                  swarn2( "Now req has: ", $req->header( "If-Modified-Since" ) );
              },
          );

        swarn does not work well this case; this has deep frames.

        Oops, that case needs more depth. Increase BIAS value these cases.

          ( my $ua = LWP::UserAgent->new )->add_handler(
              request_prepare => sub {
                  my( $req, $ua, $h ) = @_;
                  local $Log::Sigil::BIAS += 4;
                  swarn( "Adding" );
              },
          );
          $ua->get( "http://example.com/" );
          # --> +++ Adding by .../LWP/UserAgent.pm[243]: LWP::UserAgent::prepare_request

PROPERTIES
    SIGILS
        Is a array which are used as prefix.

    TIMES
        Specifies how many sigil is repeated.

    SEPARATOR
        Will be placed between sigils and log message.

    BIAS
        Controls caller frame depth.

AUTHOR
    kuniyoshi kouji <kuniyoshi@cpan.org>

SEE ALSO
LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Log::Sigil

1.01    Sun Mar 17 23:11:14 JST 2013
        - tidy up module, use only one function that is swarn.
0.15    Sat Aug 20 14:26:52 2011
        - Need perl 5.8.0 or later
0.14    Sat Aug 13 13:18:10 2011
        - autodie joins corelist at 5.10, so i use or die instead.
0.13    Sat Aug 13  9:34:10 2011
        - Fix a bug that makes hash-ref on Readonly.
0.12    Sat Aug  6 16:35:00 2011
        - Change test using from IO::Scalar to open.
0.11    foo bar baz 2011
        - Fix a bug that ignores new's arguments.
        - Add bias, and quiet attributes.
        - Add sayf, and warnf methods.
0.10    Sun Jun 12 19:57:00 2011
        - Fix a bug that prints invalid warnings suffix.
          The suffix likes 'foo at main.pl line 8.'.
0.09    Thu Jun  7 23:28:00 2011
        - Add requirement module to Makefile.PL, it is Test::Output
0.08    Sat Jun  4 18:28:00 2011
        - 'new' returns instance but carps.
        - Fix a some bugs, but i don't remember those.
0.07    Sun May 16 00:33:00 2011
        - Fix a 'requires' bug in Makefile.PL.
0.06    Sun May 16 00:27:00 2011
        - Suppot subroutine level change.  When caller subroutine changes,
          this repots it as reset sigil.
        - new will be croak, use 'instance' instead.
0.05    Sun May 15 23:44:00 2011
        - Fix a bug in multi messages, that prints odd suffix.
        - Supports $, var.
0.04    Sun May 15 00:21:00 2011
        - Fix a bug in multi messages, that prints all messages are formated.
          Needs only first message is formated.
0.03    Sat May 14 23:55:00 2011
        - Supports more than one message in say, warn, and dump.
0.02    Sat May 14 22:50:00 2011
        - Add a dump method
0.01    Fri Apr 29 04:33:57 2011
        - original version


sample/sam.pl
#!/usr/bin/perl
use 5.10.0;
use utf8;
use strict;
use warnings;
use open qw( :utf8 :std );
use FindBin;
use lib "$FindBin::Bin/../lib";
use Log::Sigil;

my $log = Log::Sigil->instance;

$log->sigils( q{#}, qw( - + ) );

$, = "T";
#$log->warn( qw( foo bar baz ) );

package Foo;
sub foo {
    $log->warn( "Foo::foo" );
    $log->warn( "Foo::foo" );
    $log->warn( "Foo::foo" );
}

sub bar {
    $log->warn( "Foo::bar" );
    $log->warn( "Foo::bar" );
    $log->warn( "Foo::bar" );
}

package Bar;
sub foo {
    $log->warn( "Bar::foo" );
    $log->warn( "Bar::foo" );
    $log->warn( "Bar::foo" );
}

sub bar {
    $log->warn( "Bar::bar" );
    $log->warn( "Bar::bar" );
    $log->warn( "Bar::bar" );
}

package main;
$log->warn( "foo" );
$log->warn( "foo" );

local $, = "TTT";

Foo::foo();
Bar::foo();
Foo::bar();
Bar::foo();
Bar::bar();

$log->warn( "foo" );
$log->warn( "foo" );

$log->dump( [ ] );



lib/Log/Sigil.pm
package Log::Sigil;
use strict;
use warnings;
use Exporter "import";
use List::Util qw( max );

use constant DEBUG => 0;

our @EXPORT  = qw( swarn swarn2 );
our $VERSION = "1.01";

our @SIGILS    = (
    qw(
        =
        +
        !
        @
    ),
    q{#},
    qw(
        $
        %
        ^
        &
        *
        -
        |
        \
        ~
        ?
    ),
);
our $TIMES     = 3;
our $SEPARATOR = q{ };
our $BIAS      = 0;
our %INDEX     = ( "main::" => 0 ); # Ensure `values` + 1 is the next.
my $ANON_REGEX = qr{ (?: .*::__ANON__ | [(]eval[)] ) \z}msx;

sub swarn {
    my $nth  = 0;
    my $bias = 1;

    $nth++
        while caller $nth;

    my( $package, $filename, $line, $subroutine ) = caller $nth - $bias - $BIAS;

    $bias++;

    $subroutine = "main::"
        if $subroutine eq join q{::}, __PACKAGE__, "swarn";

    $subroutine = "${subroutine}::$line"
        if $subroutine =~ m{$ANON_REGEX};

    $bias++
        if $subroutine =~ m{$ANON_REGEX};

    if ( my @list = caller $nth - $bias - $BIAS ) {
        ( undef, undef, $line ) = @list;
    }

warn "\$package:\t$package"       if DEBUG;
warn "\$filename:\t$filename"     if DEBUG;
warn "\$line:\t$line"             if DEBUG;
warn "\$subroutine:\t$subroutine" if DEBUG;

    unless ( exists $INDEX{ $subroutine } ) {
        $INDEX{ $subroutine } = max( values %INDEX ) + 1;
    }

    my $sigil = $SIGILS[ $INDEX{ $subroutine } % @SIGILS ];
warn "\$sigil:\t$sigil" if DEBUG;
    unshift @_, $sigil x $TIMES, $SEPARATOR;
    push @_, " by ${filename}[$line]: $subroutine\n"; # Ignore if original has \n at the end.

    warn @_;
}

sub swarn2 {
    local $BIAS = $BIAS + 1;
    &swarn;
}

1;
__END__

=head1 NAME

Log::Sigil - show warnings with sigil prefix

=head1 SYNOPSIS

  filename: synopsis.pl
   1 use Log::Sigil qw( swarn swarn2 );
   2
   3 sub foo {
   4     swarn( "foo" );
   5     swarn( "bar" );
   6 }
   7
   8 swarn( "foo" );
   9
  10 foo( );
  11
  12 swarn( "bar" );

  above shows:
  === foo by synopsis.pl[8]: main::
  +++ foo by synopsis.pl[4]: main::foo
  +++ bar by synopsis.pl[5]: main::foo
  === bar by synopsis.pl[12]: main::

=head1 DESCRIPTION

This module helps printing debug by adding prefix to the warning message.
The prefix will change if caller changes, meaning 'foo' sub, and 'bar' sub
have different prefix each other.

i do printing debug frequently.  In debugging, my warning messages became
too big to read.  When i in trouble (yes, so doing printing
debug), i do not want to remove the warning messages.  Once i thought it is needed,
it is needed twice, and more.  Thus, i need a format which can read warning messages
even if that is big.

=head1 EXPORTS

=over

=item swarn

=item swarn2

=back

=head1 FUNCTIONS

=over

=item swarn

Works all of this module does.  That are,
adding prefix,
setting up filename and line,
and, setting up package and subroutine.

=item swarn2

Same as swarn, but has a 1 bias.  This is useful when calling from
some handler subroutine, such as;

  ( my $ua = LWP::UserAgent->new )->add_handler(
      request_prepare => sub {
          my( $req, $ua, $h ) = @_;
          swarn2( "Adding If-Modified-Since..." );
          $req->...
          swarn2( "Now req has: ", $req->header( "If-Modified-Since" ) );
      },
  );

swarn does not work well this case; this has deep frames.

Oops, that case needs more depth.  Increase BIAS value these cases.

  ( my $ua = LWP::UserAgent->new )->add_handler(
      request_prepare => sub {
          my( $req, $ua, $h ) = @_;
          local $Log::Sigil::BIAS += 4;
          swarn( "Adding" );
      },
  );
  $ua->get( "http://example.com/" );
  # --> +++ Adding by .../LWP/UserAgent.pm[243]: LWP::UserAgent::prepare_request

=back

=head1 PROPERTIES

=over

=item SIGILS

Is a array which are used as prefix.

=item TIMES

Specifies how many sigil is repeated.

=item SEPARATOR

Will be placed between sigils and log message.

=item BIAS

Controls caller frame depth.

=back

=head1 AUTHOR

kuniyoshi kouji E<lt>kuniyoshi@cpan.orgE<gt>

=head1 SEE ALSO

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

